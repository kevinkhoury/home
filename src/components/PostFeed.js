import React from 'react'
import PropTypes from 'prop-types'
import PostCard from './PostCard'
import { v4 as uuidv4 } from 'uuid'
import moment from 'moment'
import { addTrailingSlash } from '../utils'
import { featuredImagePropTypes } from '../proptypes'

const Moment = moment().constructor

const PostFeed = ({ isPreview, posts }) => (
  <div className="post-feed">
    {!isPreview &&
      !!posts &&
      !!posts.length &&
      posts.map(({ image, slug, pageTitle, date }, index) => {
        return (
          <PostCard
            key={uuidv4()}
            count={index}
            image={image}
            slug={addTrailingSlash(slug)}
            pageTitle={pageTitle}
            date={date}
          />
        )
      })}
    {!isPreview && (!posts || !posts.length) && (
      <PostCard
        count={0}
        pageTitle="No posts yet. Please check back again soon!"
      />
    )}
    {!!isPreview && (
      <PostCard
        count={0}
        pageTitle="Your posts will appear here in reverse chronological order"
      />
    )}
  </div>
)

PostFeed.propTypes = {
  isPreview: PropTypes.bool,
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      image: featuredImagePropTypes,
      slug: PropTypes.string,
      pageTitle: PropTypes.string.isRequired,
      date: PropTypes.instanceOf(Moment),
    }),
  ),
}

export default PostFeed
