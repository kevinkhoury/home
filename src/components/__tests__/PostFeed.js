import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import moment from 'moment'
import PostFeed from '../PostFeed'

const mockPost = (id) => ({
  image: {
    src: {
      childImageSharp: {
        fluid: {},
        original: {
          height: 100,
          width: 200,
        },
      },
    },
    d: {
      childImageSharp: {
        fluid: {
          foobar: `${id} desktop fluid`,
        },
      },
    },
    m: {
      childImageSharp: { fluid: { foobar: `${id} mobile fluid` } },
    },
    alt: `${id}alt`,
    caption: `${id}caption`,
  },
  slug: `/${id}`,
  pageTitle: `${id}`,
  date: moment('20200101', 'YYYYMMDD'),
})

const DEFAULT_PROPS = {
  isPreview: false,
  posts: [mockPost('foo'), mockPost('bar'), mockPost('baz')],
}

jest.mock('../PostCard', () => (props) => (
  <div data-testid="postcard">{JSON.stringify(props)}</div>
))

jest.mock('uuid', () => ({
  v4: jest
    .fn(() => {
      const { v4 } = jest.requireActual('uuid')
      return v4()
    })
    .mockImplementationOnce(() => 'uuid1')
    .mockImplementationOnce(() => 'uuid2')
    .mockImplementationOnce(() => 'uuid3'),
}))

afterEach(() => {
  jest.clearAllMocks()
})

describe('PostFeed', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<PostFeed {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('generates a card for each post', () => {
    render(<PostFeed {...DEFAULT_PROPS} />)
    const foo = screen.getByText(/"alt":"fooalt"/)
    const bar = screen.getByText(/"alt":"baralt"/)
    const baz = screen.getByText(/"alt":"bazalt"/)
    expect(foo).toBeTruthy()
    expect(bar).toBeTruthy()
    expect(baz).toBeTruthy()
  })
  it('generates a placeholder during preview', () => {
    const props = { ...DEFAULT_PROPS, isPreview: true }
    render(<PostFeed {...props} />)
    const result = screen.getByText(
      /Your posts will appear here in reverse chronological order/,
    )
    expect(result).toBeTruthy()
  })
  it('generates a placeholder when there are no posts', () => {
    const props = { ...DEFAULT_PROPS, posts: [] }
    render(<PostFeed {...props} />)
    const result = screen.getByText(
      /No posts yet\. Please check back again soon!/,
    )
    expect(result).toBeTruthy()
  })
})
