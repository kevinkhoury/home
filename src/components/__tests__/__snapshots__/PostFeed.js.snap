// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostFeed snapshot renders properly 1`] = `
<div
  className="post-feed"
>
  <div
    data-testid="postcard"
  >
    {"count":0,"image":{"src":{"childImageSharp":{"fluid":{},"original":{"height":100,"width":200}}},"d":{"childImageSharp":{"fluid":{"foobar":"foo desktop fluid"}}},"m":{"childImageSharp":{"fluid":{"foobar":"foo mobile fluid"}}},"alt":"fooalt","caption":"foocaption"},"slug":"/foo/","pageTitle":"foo","date":"2020-01-01T08:00:00.000Z"}
  </div>
  <div
    data-testid="postcard"
  >
    {"count":1,"image":{"src":{"childImageSharp":{"fluid":{},"original":{"height":100,"width":200}}},"d":{"childImageSharp":{"fluid":{"foobar":"bar desktop fluid"}}},"m":{"childImageSharp":{"fluid":{"foobar":"bar mobile fluid"}}},"alt":"baralt","caption":"barcaption"},"slug":"/bar/","pageTitle":"bar","date":"2020-01-01T08:00:00.000Z"}
  </div>
  <div
    data-testid="postcard"
  >
    {"count":2,"image":{"src":{"childImageSharp":{"fluid":{},"original":{"height":100,"width":200}}},"d":{"childImageSharp":{"fluid":{"foobar":"baz desktop fluid"}}},"m":{"childImageSharp":{"fluid":{"foobar":"baz mobile fluid"}}},"alt":"bazalt","caption":"bazcaption"},"slug":"/baz/","pageTitle":"baz","date":"2020-01-01T08:00:00.000Z"}
  </div>
</div>
`;
